### PR TITLE
Performance improvement of `FermionicOp.reduce`

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

This PR introduces the following two changes.
- Vectorize `FerimionicOp.reduce`
- Add `normal_order` option to `FerimionicOp.reduce` (default is True)

```python
import random
from timeit import timeit
from qiskit_nature.operators.second_quantization import FermionicOp


def random_op(k, n):
    random.seed(123)
    op = sum((FermionicOp("".join(random.choices("+-ENI", k=k)), display_format="dense") * random.random()) for _ in range(n))
    return op

def bench_reduce(op):
    print(f"normal order: {timeit(lambda: op.reduce(), number=1)} sec")
    try:
        print(f"no normal order: {timeit(lambda: op.reduce(normal_order=False), number=1)} sec")
    except TypeError:
        pass

op = random_op(5, 10000)
bench_reduce(op)
```

main
```
normal order: 2.4304106099998535 sec
```

this PR
```
normal order: 2.271573748999799 sec
no normal order: 0.05692009299946221 sec
```

### Details and comments

The motivation to apply `reduce` before qubit mapping is #585. The output operator can be redundant.
So, I want to simplify the operators. But, I don't think qubit operator requires the normal order.

I added the `normal_order` option because applying `normal_order` may increase the number of operators and `to_normal_order` can be slow.
If users apply `reduce` to an operator before qubit mapping, it takes more time than without `reduce` due to the larger number of operators.
I attach an example. The size of operator grows from 1000 to 18950 by applying `reduce`.

```python
import random
from timeit import timeit

from qiskit_nature.operators.second_quantization import FermionicOp
from qiskit_nature.converters.second_quantization import QubitConverter
from qiskit_nature.mappers.second_quantization import JordanWignerMapper

def random_op(k, n):
    random.seed(123)
    op = sum((FermionicOp("".join(random.choices("+-E", k=k)), display_format="dense") * random.random()) for _ in range(n))
    return op

qubit_converter = QubitConverter(mapper=JordanWignerMapper())

def bench_convert(op):
    print(f"w/o reduce: {timeit(lambda: qubit_converter.convert(op), number=1)} sec")
    print(f"w/ reduce of normal order: {timeit(lambda: qubit_converter.convert(op.reduce()), number=1)} sec")
    try:
        print(f"w/ reduce no normal order: {timeit(lambda: qubit_converter.convert(op.reduce(normal_order=False)), number=1)} sec")
    except TypeError:
        pass

op = random_op(10, 1000)

print("qubit convert")
bench_convert(op)

print("\noperator sizes")
print(f"original op: {len(op)}")
print(f"w/ reduce of normal order: {len(op.reduce())}")
try:
    print(f"w/ reduce no normal order: {len(op.reduce(normal_order=False))}")
except TypeError:
    pass
```

this PR
```
qubit convert
w/o reduce: 1.7863747950004836 sec
w/ reduce of normal order: 17.12350188399978 sec
w/ reduce no normal order: 1.741173968000112 sec

operator sizes
original op: 1000
w/ reduce of normal order: 18950
w/ reduce no normal order: 989
```

### CODE REVIEW

**Comment by t-imamichi** [comment]
I have not written the docstring of the new option `reduce(normal_order)`. If the option is allowed, I will write it.

File: qiskit_nature/operators/second_quantization/fermionic_op.py [hunk 1]
```diff
@@ -538,24 +538,25 @@ def adjoint(self) -> FermionicOp:
             display_format=self.display_format,
         )
 
-    def reduce(self, atol: Optional[float] = None, rtol: Optional[float] = None) -> FermionicOp:
+    def reduce(
```
Comment (pbark) [inline qiskit_nature/operators/second_quantization/fermionic_op.py:541]: Why we do not have a docstring here?

**Comment by kevinsung** [comment]
This PR is made obsolete by https://github.com/Qiskit/qiskit-nature/pull/607.

**Comment by t-imamichi** [comment]
I will port this optimization to the upcoming `simplify` methods #607. 

https://github.com/Qiskit/qiskit-nature/pull/607#issuecomment-1076457920

**Pull request closed without merging**